### PR TITLE
Core working with upstream/downstream machine and status sending tested with mbc-playout.

### PR DESCRIPTION
**Pull request by fabriciocosta**
Please if make doesnt work, try it running melted and setting root of melted with the directory of you mbc-mosto rep..
You can change the images/black.png with you favorite image.
run melted:
melted
telnet localhost 5250
UADD sdl
SET root=/yourdirectoryto/mbc-mosto

then in mbc-mosto
node server.js

then run mbc-playout with:
node server.js

### CODE REVIEW

**Comment by Lacrymology** [comment]
There's a lot of commits that aren't in master here, notice

```
Lacrymology     start preparing the CaspaDriver object …  031cd48
Lacrymology     use mbc-common >1.x's pubsub driver …     4458b70
Lacrymology     return a CaspaDriver object …     d4e5d1b
Lacrymology     keep the mubsub syntax …  07a7050
Lacrymology     this has various channels …   412dfba
Lacrymology     add publishStatus-specific method …   fd50440
Lacrymology     _id is always 1 …     a7e2b95
Lacrymology     make CaspaDriver into an EventEmitter …   5bfe579
xaiki   require MELTED for melted-run …   3a3a7c3
Lacrymology     Prepare to send 'ready' message when all setup is done …  8d653a0
Lacrymology     set up status collection …    67dc9c8
Lacrymology     make these local vars …   b8e4e90
Lacrymology     default id is 2 …     eb6e5a5
Lacrymology     udpate Status defaults …  64a7706
Lacrymology     `publish` is gone …   a7c44ee
Lacrymology     move setup functions to a function of it's own, it's better for defin… …    952a3f1
Lacrymology     bugfix …  c6da472
Lacrymology     convert and populate status to publish …  7838b46
Lacrymology     comments …    b90edf8
Lacrymology     message class. This should very probably be a Backbone model, and mov… …    fa58c97
Lacrymology     drop any sticky errors that might be left from the last run on setup …    d782be4
Lacrymology     publish message …     184c5f4
Lacrymology     drop sticky message …     ae6556a
Lacrymology     percentage will be float 0..1 …   e3788c9
Lacrymology     @jmrunge will provide me with actual previous, current, next Playlists …  fd5e4cb 
```

these are probably from a merge master of a reverted commit (probably @xaiki's one)

File: drivers/playlists/mongo-driver.js [hunk 1]
```diff
@@ -200,7 +200,7 @@ function mongo_driver(conf) {
                 medias.push(new Media(block_id, orig_order, actual_order, playlist_id, clip_name, type, file, length, parseFloat(fps)));
             });
 
-            var playlist = new Playlist(playlist_id, name, startDate, medias, endDate);
+            var playlist = new Playlist(playlist_id, name, startDate, medias, endDate, "snap");
```
Comment (jmrunge) [inline drivers/playlists/mongo-driver.js:203]: There should be a parameter in playlist driver config to define how should we treat a playlist that comes without this attribute.

File: drivers/status/pubsub.js [hunk 2]
```diff
@@ -83,18 +83,35 @@ function CaspaDriver() {
                 _id: clip.id
             }
         }
+
         var clips = meltedStatus.clips;
+        var p_clip,c_clip,n_clip;
+        
+        if (0<=meltedStatus.clip.previous && meltedStatus.clip.previous<clips.length) {
+			p_clip = makePiece(clips[meltedStatus.clip.previous]);
```
Comment (jmrunge) [inline drivers/status/pubsub.js]: Indent?

File: drivers/status/pubsub.js [hunk 3]
```diff
@@ -83,18 +83,35 @@ function CaspaDriver() {
...
+        } else p_clip = { name: '' };
+
+        if (0<=meltedStatus.clip.current && meltedStatus.clip.current<clips.length) {
+			c_clip = makePiece(clips[meltedStatus.clip.current]);
```
Comment (jmrunge) [inline drivers/status/pubsub.js]: Indent?

File: drivers/status/pubsub.js [hunk 4]
```diff
@@ -83,18 +83,35 @@ function CaspaDriver() {
...
+        } else c_clip = { name: '', progress: '0%' };
+
+        if (0<=meltedStatus.clip.next && meltedStatus.clip.next<clips.length) {
+			n_clip = makePiece(clips[meltedStatus.clip.next]);
```
Comment (jmrunge) [inline drivers/status/pubsub.js]: Indent?

File: drivers/status/pubsub.js [hunk 5]
```diff
@@ -83,18 +83,35 @@ function CaspaDriver() {
...
+        } else n_clip = { name: '' };
+
         var status = {
+            _id: 2,
```
Comment (jmrunge) [inline drivers/status/pubsub.js]: Why is this hardcoded?

Comment (Lacrymology) [inline drivers/status/pubsub.js]: there's only 1 status object on the DB, and that's it's ID.

It's in caspa/models/App.Status

This'll be migrated to mbc-common (to use the backbone model) soon

Comment (jmrunge) [inline drivers/status/pubsub.js]: Ok.  We should try to make sure this is the _id used everywhere.  Once
migrated to mbc-common, can we make it a constant?

2013/4/18 Tomas Neme notifications@github.com

Comment (Lacrymology) [inline drivers/status/pubsub.js]: what ID? you should not be making IDs anywhere in your part of the status handling. This is just for the mosto->caspa communication, only this driver should care about this

Comment (jmrunge) [inline drivers/status/pubsub.js]: Well, this is INSIDE status driver...

2013/4/18 Tomas Neme notifications@github.com

File: mosto.js [hunk 6]
```diff
@@ -1,24 +1,55 @@
...
+util             = require('util'),
+events           = require('events'),
+moment           = require('moment'),
+Playlist         = require('./api/Playlist'),
+Media            = require('./api/Media'),
+mvcp_server      = require('./drivers/mvcp/mvcp-driver'),
+playlists_driver = require('./drivers/playlists/playlists-driver'),
+status_driver    = require('./drivers/status/pubsub'),
+config           = require('./drivers/mvcp/conf/melted-node-driver'),
+utils            = require('./utils');
```
Comment (jmrunge) [inline mosto.js]: Indent?

File: drivers/status/pubsub.js (hunk 5 above)
Comment (Lacrymology) [inline drivers/status/pubsub.js]: I know, I mean I don't understand why do you say 

> we should try to make sure this is the _id used everywhere

I don't think I ever override the ID anywhere here..

File: mosto.js [hunk 7]
```diff
@@ -1,24 +1,55 @@
...
+var fs           = require('fs'),
+util             = require('util'),
+events           = require('events'),
+moment           = require('moment'),
+Playlist         = require('./api/Playlist'),
+Media            = require('./api/Media'),
+mvcp_server      = require('./drivers/mvcp/mvcp-driver'),
+playlists_driver = require('./drivers/playlists/playlists-driver'),
+status_driver    = require('./drivers/status/pubsub'),
+config           = require('./drivers/mvcp/conf/melted-node-driver'),
```
Comment (jmrunge) [inline mosto.js]: Why are you requiring melted-node-driver conf??? It should be used only by melted-node-driver!

File: mosto.js [hunk 8]
```diff
@@ -1,24 +1,55 @@
...
+
+
+
+function ScheduledMedia( media, schedule_time, schedule_duration, expected_start, expected_end ) {
+    this.media   = media;
+    this.schedule_time = schedule_time;
+    this.schedule_duration = schedule_duration;
+    this.expected_start = expected_start;
+    this.expected_end = expected_end;
+};
```
Comment (jmrunge) [inline mosto.js]: Shouldnt this be moved to API?

File: drivers/status/pubsub.js (hunk 5 above)
Comment (jmrunge) [inline drivers/status/pubsub.js]: Ok, if that's the case then leave it this way.  I thought from the other
side of pubsub we might be looking for this id... Or something like that...
And I have something against hardcoded values! :)

2013/4/18 Tomas Neme notifications@github.com

File: mosto.js [hunk 9]
```diff
@@ -1,24 +1,55 @@
...
+        console.log("mbc-mosto: [INFO] Added playlist:\nid: " + playlist.id
+                    + "\nname: " + playlist.name
+                    + "\nstartDate: " + playlist.startDate
+                    + "\nendDate: " + playlist.endDate);
         self.orderPlaylists();
     };
     
+    /** updatePlaylist
+     *
+     *       playlist.name is the playlist key or id!
```
Comment (jmrunge) [inline mosto.js]: Please remove this line of comment, we use ids now...

File: drivers/status/pubsub.js (hunk 5 above)
Comment (fabriciocosta) [inline drivers/status/pubsub.js]: I dont fully understand this status format (id? and source?). Please
explain and enlight-me on how this reflects afterward on the caspa
interface.

2013/4/18 Juan Martin Runge notifications@github.com

> In drivers/status/pubsub.js:
> 
> > ```
> >      var status = {
> > ```
> > -            _id: 2,
> 
> Ok, if that's the case then leave it this way. I thought from the other
> side of pubsub we might be looking for this id... Or something like that...
> And I have something against hardcoded values! :) 2013/4/18 Tomas Neme <
> notifications@github.com>
> … <#13e1e66d0a4fc98f_>
>  In drivers/status/pubsub.js: > var status = { > + _id: 2, I know, I mean
> I don't understand why do you say we should try to make sure this is the
> _id used everywhere I don't think I ever override the ID anywhere here.. —
> Reply to this email directly or view it on GitHub<
> https://github.com/inaes-tic/mbc-mosto/pull/69/files#r3859988> .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/inaes-tic/mbc-mosto/pull/69/files#r3860136
> .

## 

_Fabricio Costa_
Moldeo Interactive Coop. Trab. Ltda.
www.moldeointeractive.com.ar
Castillo 1336 Módulo 12 A - C.A.B.A. - Argentina
(+54911) 6124-4201
fabricio.costa@moldeointeractive.com.ar

File: mosto.js (hunk 9 above)
Comment (fabriciocosta) [inline mosto.js]: ok

2013/4/18 Juan Martin Runge notifications@github.com

> In mosto.js:
> 
> > ```
> >      self.orderPlaylists();
> >  };
> > ```
> > -    /*\* updatePlaylist
> > -     *
> > -     \*       playlist.name is the playlist key or id!
> 
> Please remove this line of comment, we use ids now...
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/inaes-tic/mbc-mosto/pull/69/files#r3860186
> .

## 

_Fabricio Costa_
Moldeo Interactive Coop. Trab. Ltda.
www.moldeointeractive.com.ar
Castillo 1336 Módulo 12 A - C.A.B.A. - Argentina
(+54911) 6124-4201
fabricio.costa@moldeointeractive.com.ar

File: mosto.js [hunk 10]
```diff
@@ -27,16 +58,30 @@ function mosto(configFile) {
...
-                + "\nname: " + playlist.name 
-                + "\nstartDate: " + playlist.startDate 
-                + "\nendDate: " + playlist.endDate);
+        
+        //update may create if needed
+        //TODO: chech time window: startDate-endDate
+        if (i==-1) {
+            self.playlists.push(playlist);
+        } else {
+            playlist.loaded = self.playlists[i].loaded;
```
Comment (jmrunge) [inline mosto.js]: Is 'loaded' attr usefull anymore? I created it as a workaround and expected to remove it when your code arrives...

File: drivers/status/pubsub.js (hunk 5 above)
Comment (Lacrymology) [inline drivers/status/pubsub.js]: `// copied from caspa/models.App.Status`

It's the defaults for a backbone model. The _id is to make sure only a single copy of the status exists in the database at any time.

Source, I don't know, it was there. It's probably something from airtime. @xaiki?

File: mosto.js [hunk 11]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+		self.time_window_to.add( moment.duration({ hours: 4 }) );
+
+		console.log("mbc-mosto: [INFO] [FETCH] checkoutPlaylists > from: " + self.time_window_from.format("DD/MM/YYYY HH:mm:ss") + " to:" + self.time_window_to.format("DD/MM/YYYY HH:mm:ss") );
+
+		//now we read playlists, between "last_time_window_to" and "time_window_to"
+        self.driver.getPlaylists( { from: last_time_window_to, to: self.time_window_to, setWindow: false }, function(playlists) { 
+			
+			//just import new ones....
+			console.log("mbc-mosto: [INFO] [LOGIC] checkoutPlaylists > receive playlists from: " + last_time_window_to.format("DD/MM/YYYY HH:mm:ss") + " to:" + self.time_window_to.format("DD/MM/YYYY HH:mm:ss") );
+            for( var p=0;p<playlists.length;p++) {
```
Comment (jmrunge) [inline mosto.js:142]: Shouldnt you be using self.playlists instead of just playlists?

File: mosto.js (hunk 7 above)
Comment (fabriciocosta) [inline mosto.js]: i dont need it... maybe it came from another branch.

File: mosto.js (hunk 8 above)
Comment (fabriciocosta) [inline mosto.js]: Yes it would be nice, are you ok with that?

File: mosto.js [hunk 12]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+				var playlist = playlists[p];
+    		    var i = -1;
+	    	    self.playlists.some( function(element, index, array) {
+		            if (element.id === playlist.id) {
+     		            i = index;
+	    	            return true;
+		            }
+		        });
+    			if (i==-1) {
+    				self.playlists.push(playlist);
```
Comment (jmrunge) [inline mosto.js:152]: Cant you call addPlaylist here?  If we add some validation code before adding playlists, we should do it only in one place...

File: mosto.js (hunk 10 above)
Comment (fabriciocosta) [inline mosto.js]: I dont use it anymore. Remove.

File: mosto.js (hunk 7 above)
Comment (jmrunge) [inline mosto.js]: Please remove it then.

2013/4/18 Fabricio Costa Alisedo notifications@github.com

File: mosto.js (hunk 8 above)
Comment (jmrunge) [inline mosto.js]: Please move it to its own file inside api.  Thanks!

2013/4/18 Fabricio Costa Alisedo notifications@github.com

File: mosto.js (hunk 10 above)
Comment (jmrunge) [inline mosto.js]: Can yo do it please? So we keep all this changes in your branch.

2013/4/18 Fabricio Costa Alisedo notifications@github.com

File: mosto.js [hunk 13]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+		} );
+
+        //TODO: If we are up to date, just return!! (important to avoid infinite recursions)        
+        
+    }
+    
+    
+    mosto.prototype.convertFramesToSeconds = function ( frames, fps ) {
+        return frames/fps;
+    }
```
Comment (jmrunge) [inline mosto.js]: Please write this function in utils (I think there is one already!)

File: mosto.js [hunk 14]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+
+    mosto.prototype.convertLengthToMilliseconds = function ( frames ) {
+        var m = moment( frames, "HH:mm:ss.SS");
+        return m.hours()*60*60*1000 + m.minutes()*60*1000 + m.seconds()*1000 + m.milliseconds();
+    }
```
Comment (jmrunge) [inline mosto.js]: This too, to utils please.

File: mosto.js [hunk 15]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+    
+    mosto.prototype.convertFramesToMilliseconds = function ( frames, fps ) {
+		if (fps+""=="NaN" || fps==undefined || fps===false || fps==0) {
+			return	self.convertLengthToMilliseconds(frames);
+ 		}
+
+        return frames * 1000.0 / (1.0 * fps);
+    }
```
Comment (jmrunge) [inline mosto.js]: Utils

File: mosto.js [hunk 16]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+ 		}
+
+        return frames * 1000.0 / (1.0 * fps);
+    }       
+    
+    mosto.prototype.convertUnixToDate =  function ( unix_timestamp ) {
+        //var date = new Date(unix_timestamp*1000);     
+        var date = new moment(unix_timestamp);
+        return date.format("hh:mm:ss");
+    }
```
Comment (jmrunge) [inline mosto.js]: Utils

File: mosto.js [hunk 17]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+
+    mosto.prototype.convertDateToUnix =  function ( date_timestamp ) {
+        
+        var date = new moment(date_timestamp);
+        return date.unix();
+        
+    }
```
Comment (jmrunge) [inline mosto.js]: Utils

File: mosto.js [hunk 18]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+
+    mosto.prototype.startBlack = function( schedule_time, sch_duration, sch_expect_start, sch_expect_end ) {
+         var BlackMedia = new Media( 'black_id' /*id*/, '0' /*orig_order*/, '0'/*actual_order*/, '0' /*playlist_id*/, 'black' /*name*/, 'file' /*type*/, self.config.black, sch_duration/*length*/, ''/*fps*/ );
```
Comment (jmrunge) [inline mosto.js:171]: Please vars start with lowercase

File: mosto.js [hunk 19]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+         console.log("mbc-mosto: [INFO] [LOGIC] startBlack > media:" + BlackMedia.file + " schedule_time:" + schedule_time + " sch_duration:" + sch_duration + " sch_expect_start:" + sch_expect_start + " sch_expect_end:" + sch_expect_end + " fps?:"+BlackMedia.fps );
+         var medias = [];
+         medias.push(BlackMedia);
+         self.playlists.push( new Playlist( BlackMedia.id, BlackMedia.id, moment( sch_expect_start, "DD/MM/YYYY HH:mm:ss.SSS" ).toDate(), medias, moment( sch_expect_end, "DD/MM/YYYY HH:mm:ss.SSS" ).toDate(), "snap" ) );
+    }
+    
+    mosto.prototype.queueBlack = function( schedule_time, sch_duration, sch_expect_start, sch_expect_end ) {
+         //Media(id, orig_order, actual_order, playlist_id, name, type, file, length, fps)
+         var BlackMedia = new Media( 'black_id' /*id*/, '0' /*orig_order*/, '0'/*actual_order*/, '0' /*playlist_id*/, 'black' /*name*/, 'file' /*type*/, self.config.black, sch_duration/*length*/, ''/*fps*/ );
```
Comment (jmrunge) [inline mosto.js:180]: You could have a helper function to return BlackMedia just changing the init time... I think...

File: drivers/playlists/mongo-driver.js (hunk 1 above)
Comment (fabriciocosta) [inline drivers/playlists/mongo-driver.js:203]: Default is "snap", the other mode will be "fixed" like for making inserts and replace any current clip without modifying actual playlists. For "fixed" mode we need some overlays in Caspa, we will discuss this later, maybe tomorrow.

"snap" mode (like queue mode), just queue playlists and intend to snap. If schedule of the next one is far away from the end of the last one, then snap is not applied. "far away" means more than 10 seconds (should be whatever we want and placed in the config file) and consider the next playlist with his startDate as the mandatory new schedule time.

We have to discuss all this later, i already analyzed all cases but we have to implement what is really useful for now.

File: drivers/status/pubsub.js (hunk 2 above)
Comment (fabriciocosta) [inline drivers/status/pubsub.js]: i should remove the Tab key from my keyboard. But i can...not

File: drivers/playlists/mongo-driver.js (hunk 1 above)
Comment (jmrunge) [inline drivers/playlists/mongo-driver.js:203]: Ok, I just think we could give the user the possibility to decide what mode
will be the default one.  I like the "far away" being in config.  And dont
forget to send warnings if you detect a "far away".

Perhaps we could do a Hangout tomorrow?

2013/4/18 Fabricio Costa Alisedo notifications@github.com

File: drivers/status/pubsub.js (hunk 2 above)
Comment (jmrunge) [inline drivers/status/pubsub.js]: Use an editor where you can config "Tab=4 spaces".  I use Netbeans... and
gedit can be configured so too....

2013/4/18 Fabricio Costa Alisedo notifications@github.com

Comment (fabriciocosta) [inline drivers/status/pubsub.js]: done.

2013/4/18 Juan Martin Runge notifications@github.com

> In drivers/status/pubsub.js:
> 
> > ```
> >      var clips = meltedStatus.clips;
> > ```
> > -        var p_clip,c_clip,n_clip;
> >   +
> > -        if (0<=meltedStatus.clip.previous && meltedStatus.clip.previous<clips.length) {
> > -           p_clip = makePiece(clips[meltedStatus.clip.previous]);
> 
> Use an editor where you can config "Tab=4 spaces". I use Netbeans... and
> gedit can be configured so too.... 2013/4/18 Fabricio Costa Alisedo <
> notifications@github.com>
> … <#13e1e8150263ad80_>
>  In drivers/status/pubsub.js: > var clips = meltedStatus.clips; > + var
> p_clip,c_clip,n_clip; > + > + if (0<=meltedStatus.clip.previous &&
> meltedStatus.clip.previous<clips.length) { > + p_clip =
> makePiece(clips[meltedStatus.clip.previous]); i should remove the Tab key
> from my keyboard. But i can... — Reply to this email directly or view it on
> GitHubhttps://github.com/inaes-tic/mbc-mosto/pull/69/files#r3860664 .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/inaes-tic/mbc-mosto/pull/69/files#r3860766
> .

## 

_Fabricio Costa_
Moldeo Interactive Coop. Trab. Ltda.
www.moldeointeractive.com.ar
Castillo 1336 Módulo 12 A - C.A.B.A. - Argentina
(+54911) 6124-4201
fabricio.costa@moldeointeractive.com.ar

File: mosto.js [hunk 20]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+            return;
+        }
+
+        return self.preparePlaylist( next_playlist_id, lastTimeCode );
+    }
     
     mosto.prototype.sendStatus = function() {
         //TODO: Fabricio should replace all invocations to this function with
         //real invocations.  This is just an example
+console.log("FOFOFOFOFOFOFOF");
```
Comment (jmrunge) [inline mosto.js]: Can you remove this?

File: mosto.js [hunk 21]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
         self.server.getServerStatus(function(resp1) {
+	console.log("SAKDLSKLDJSADJASKLDASJ");
```
Comment (jmrunge) [inline mosto.js]: And this please....

Comment (Lacrymology) [inline mosto.js]: oops, sorry, my bad. I was testing stuff, I don't know why it got into the commit.

File: mosto.js [hunk 22]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+		            prevPlaylistId = self.playlists[index - 1].id;
+		        if (index < (self.playlists.length - 1))
+		            nextPlaylistId = self.playlists[index + 1].id;
 
-            if (index > 0) 
-                prevPlaylistId = self.playlists[index - 1].id;
-            if (index < (self.playlists.length - 1))
-                nextPlaylistId = self.playlists[index + 1].id;
+			}
+*/
```
Comment (jmrunge) [inline mosto.js:405]: Why did you change this?

File: mosto.js [hunk 23]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+		        prevClip = _.find(serverPlaylist, function(prevClip) {
+		            return parseInt(prevClip.order) === (parseInt(currentClip.order) - 1);
+		        });
+		    }
+		    if (parseInt(currentClip.order) < (serverPlaylist.length - 1)) {
+		        nextClip = _.find(serverPlaylist, function(nextClip) {
+		            return parseInt(nextClip.order) === (parseInt(currentClip.order) + 1);
+		        });
+		    }
+*/
```
Comment (jmrunge) [inline mosto.js:438]: Why did you change this? It worked for me...

File: mosto.js [hunk 24]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+    mosto.prototype.validatePlaylist = function( playlist ) {
+        for( i=0; i<playlist.length; i++) { 
+            element = playlist[i];
+            if ( element.fps != self.config.fps ) {                     
+                console.error("Playlist is not valid! media: " + element.file + " fps:" + element.fps );                                                                                        
+                return false;
+            }
+        };
+        return true;
+    }
```
Comment (jmrunge) [inline mosto.js]: Is this mandatory? We should only send clips with the fps defined in config file?

File: mosto.js [hunk 25]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+
+
+    mosto.prototype.convertMediaFileToXmlFile = function( media ) {
+        var root = process.cwd();
+        var fileName = utils.getXmlFileNameFromClip( media );
+        var xmlFile = root + config.playlists_xml_dir + "/" + fileName;
+        
+        return '"'+xmlFile+'"';
+    };
```
Comment (jmrunge) [inline mosto.js:465]: Why are you doing this?

File: mosto.js [hunk 26]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+
+    mosto.prototype.convertMediaFileToClipId = function( media ) {
+        return media.id;
+    };
```
Comment (jmrunge) [inline mosto.js:469]: What is this for?

File: mosto.js [hunk 27]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+
+
+            // CHECK AND COMPARE IF WE ARE PLAYING THE EXPECTED ONE...
+            console.log("COMPARE!!!" + self.actual_playing_status + " self.actual_playing_clip:"+self.actual_playing_clip + " vs expected: " + self.convertMediaFileToClipId(expected_clip.media) );
+
+			// NOW COMPARE EXPECTED WITH ACTUAL
+			// Conditions of timer_difference may trigger
+            if ( 
+				 Math.abs(self.timer_difference)<20000
+				 && self.actual_playing_clip != ""
```
Comment (jmrunge) [inline mosto.js:618]: Should this be on config too? Or a constant may be?

File: mosto.js [hunk 28]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+                else self.sync_lock = false;                            
+                }
+                self.sync_lock = false;
+
+            } else {
+                //WE ARE NOT PLAYING THE EXPECTED ONE (OR NOT A THE EXPECTED POSITION) !!!
+                // LOAD EVERYTHING STARTING FROM THE EXPECTED CLIP
+
+                //clean playlist (doesnt matter, we are stopped)... then populate...
+                self.server.clearPlaylist( function(response) {
```
Comment (jmrunge) [inline mosto.js:703]: I think that we should remove all clips but the playing clip (cleanPlaylist, not clearPlaylist)

File: mosto.js [hunk 29]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+
+            } else {
+                //WE ARE NOT PLAYING THE EXPECTED ONE (OR NOT A THE EXPECTED POSITION) !!!
+                // LOAD EVERYTHING STARTING FROM THE EXPECTED CLIP
+
+                //clean playlist (doesnt matter, we are stopped)... then populate...
+                self.server.clearPlaylist( function(response) {                                         
+                    console.log("clearPlaylist ok! " + response ); 
+                    sched_clip = self.scheduled_clips[self.cursor_scheduled_clip];
+                    self.server.loadClip( sched_clip.media, function(response) {
```
Comment (jmrunge) [inline mosto.js:706]: Here, I will do append instead of load

File: mosto.js [hunk 30]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+                        //EASY only C from CRUD
+                        self.server.play(  function(response) {
```
Comment (jmrunge) [inline mosto.js:708]: And here a goto [expected frame of clip appended], 1 (order of clip, since there's only two clips loaded, the appended one is order 1)

Comment (jmrunge) [inline mosto.js:708]: And then remove the old clip. :)

File: mosto.js [hunk 31]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+    previous = Playlist
+    current = Playlist
+    next = Playlist
+  }
+  position = playback % of current
+  clips = [clip list]
+};
+*/
+		//self.status_driver.setStatus( meltedStatus );
+
```
Comment (jmrunge) [inline mosto.js]: Can you remove all this comments?

File: mosto.js [hunk 32]
```diff
@@ -195,53 +997,112 @@ function mosto(configFile) {
...
+    this.config     = false;
     this.server_started = false;
+    this.default_config = {
+        "fps": "30",
+        "resolution": "hd",
+        "playout_mode": "direct",
+        "playlists_maxlength": "24:00:00",
+        "scheduled_playlist_maxlength": "04:00:00",
+        "timer_interval": "00:00:00"
+    };
```
Comment (jmrunge) [inline mosto.js]: This should be written in config only, doesnt it?

File: mosto.js [hunk 33]
```diff
@@ -195,53 +997,112 @@ function mosto(configFile) {
...
-    this.server     = new mvcp_server(this.config.mvcp_server);
-    this.driver     = new playlists_driver(this.config.playlist_server);
+    this.config = require(this.configFile);     
     
     console.log("mbc-mosto: [INFO] Starting mbc-mosto... ") ;
     
-    self.startMvcpServer(self.playPlaylists);
+    this.server     = new mvcp_server(this.config.mvcp_server);
+    this.driver     = new playlists_driver(this.config.playlist_server);
+	this.status_driver = status_driver();
```
Comment (jmrunge) [inline mosto.js:983]: Status driver should be configured in config too...

File: mosto.js (hunk 32 above)
Comment (fabriciocosta) [inline mosto.js]: Yes. To all... cleaning code...

File: mosto.js [hunk 34]
```diff
@@ -66,126 +117,877 @@ function mosto(configFile) {
...
+            if ( pl.mode == "snap") {
+                // just add all clips with schedule: "now" if diff is less than 10 seconds, else push with own timecode (sch_time = sch_expect_start
+				is = 0;
+				sch_expect_end = moment( pl_tc, "DD/MM/YYYY HH:mm:ss").format("DD/MM/YYYY HH:mm:ss.SSS");
+                for( i=0; i<pl.medias.length; i++) {
+					var sMedia = pl.medias[i];
+                    milis = self.convertFramesToMilliseconds( sMedia.length, sMedia.fps );
+		        	sch_duration_m = moment.duration( { milliseconds: milis } );        			
+
+                    sch_time = moment( sch_expect_end, "DD/MM/YYYY HH:mm:ss.SSS").format("DD/MM/YYYY HH:mm:ss.SSS");
```
Comment (jmrunge) [inline mosto.js:286]: This variable without declaration (var keyword) makes mocha fail with a global leak detected...

**Comment by jmrunge** [comment]
Please clean up the code and fix/responde issues and push so I can continue reviewing.  I would like to merge this ASAP.

File: mosto.js (hunk 11 above)
Comment (fabriciocosta) [inline mosto.js:142]: No, this is the callback of getPlaylists who receive a playlists object that we check with our actual self.playlists....

File: mosto.js (hunk 13 above)
Comment (fabriciocosta) [inline mosto.js]: Done

File: mosto.js (hunk 14 above)
Comment (fabriciocosta) [inline mosto.js]: Done

File: mosto.js (hunk 15 above)
Comment (fabriciocosta) [inline mosto.js]: done

File: mosto.js (hunk 16 above)
Comment (fabriciocosta) [inline mosto.js]: done

File: mosto.js (hunk 17 above)
Comment (fabriciocosta) [inline mosto.js]: done

File: mosto.js (hunk 18 above)
Comment (fabriciocosta) [inline mosto.js:171]: done

File: mosto.js (hunk 20 above)
Comment (fabriciocosta) [inline mosto.js]: done

File: mosto.js (hunk 21 above)
Comment (fabriciocosta) [inline mosto.js]: done

File: mosto.js (hunk 22 above)
Comment (fabriciocosta) [inline mosto.js:405]: What if index is -2 ?

File: mosto.js (hunk 23 above)
Comment (fabriciocosta) [inline mosto.js:438]: It breaks for me if serverPlaylist is undefined (node v0.10.3), i think that was why.

File: mosto.js (hunk 24 above)
Comment (fabriciocosta) [inline mosto.js]: I dont use it anymore... because i dont have any fps information from the medias loaded by Caspa... so?

File: mosto.js (hunk 25 above)
Comment (fabriciocosta) [inline mosto.js:465]: Dont use it anymore, removing.

File: mosto.js (hunk 26 above)
Comment (fabriciocosta) [inline mosto.js:469]: Inheritance from last function comparing medias from db and medias from melted.

File: mosto.js (hunk 11 above)
Comment (jmrunge) [inline mosto.js:142]: Ok! got confused with by the indentation... :)

File: mosto.js (hunk 27 above)
Comment (fabriciocosta) [inline mosto.js:618]: Yes, not fully sure how to name it: "reload_trigger_timer_difference: 20000 ms", if what is playing right now is not in sync with schedule expected, reload the clips and maybe even position the frame where it should be.

Comment (fabriciocosta) [inline mosto.js:618]: Now its self.config.reload_timer_diff

File: mosto.js (hunk 28 above)
Comment (fabriciocosta) [inline mosto.js:703]: Can we start this discussion elsewhere?

File: mosto.js (hunk 22 above)
Comment (jmrunge) [inline mosto.js:405]: How could index be -2? Im ok with the validation of line 419, but besides that I only see you removed underscore functions and used a for next block.  Didnt underscore worked for you?

File: mosto.js (hunk 23 above)
Comment (jmrunge) [inline mosto.js:438]: Again, ok with the validation, but why did you remove underscore functions?

File: mosto.js (hunk 24 above)
Comment (jmrunge) [inline mosto.js]: We are getting empty fps info from Caspa????? @Lacrymology ??

File: mosto.js (hunk 26 above)
Comment (jmrunge) [inline mosto.js:469]: Ok, are you removig it too?

File: mosto.js (hunk 29 above)
Comment (fabriciocosta) [inline mosto.js:706]: Ok, its the same discussion. Do we interrupt the actual play? It can be improved, i didn't have implemented the "goto" feature yet.

File: mosto.js (hunk 31 above)
Comment (fabriciocosta) [inline mosto.js]: done

File: mosto.js (hunk 27 above)
Comment (jmrunge) [inline mosto.js:618]: Ok!

File: mosto.js (hunk 34 above)
Comment (fabriciocosta) [inline mosto.js:286]: done

File: drivers/playlists/mongo-driver.js (hunk 1 above)
Comment (jmrunge) [inline drivers/playlists/mongo-driver.js:203]: I will fill an issue with this and assign it to @Lacrymology

File: mosto.js (hunk 12 above)
Comment (jmrunge) [inline mosto.js:152]: Filled issue to @fabriciocosta

File: mosto.js (hunk 19 above)
Comment (jmrunge) [inline mosto.js:180]: Issue to @fabriciocosta

File: mosto.js (hunk 24 above)
Comment (Lacrymology) [inline mosto.js]: it depends on the media file, I copy what comes from the DB.... sometimes ffmpeg doesn't know the files' FPS...

Comment (xaiki) [inline mosto.js]: but melted does,
and so does melt.

On 19 April 2013 17:34, Tomas Neme notifications@github.com wrote:

> In mosto.js:
> 
> > ```
> >      return status;
> >  };
> > ```
> > -    mosto.prototype.validatePlaylist = function( playlist ) {
> > -        for( i=0; i<playlist.length; i++) {
> > -            element = playlist[i];
> > -            if ( element.fps != self.config.fps ) {
> > -                console.error("Playlist is not valid! media: " + element.file + " fps:" + element.fps );
> > -                return false;
> > -            }
> > -        };
> > -        return true;
> > -    }
> 
> it depends on the media file, I copy what comes from the DB.... sometimes
> ffmpeg doesn't know the files' FPS...
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/inaes-tic/mbc-mosto/pull/69/files#r3881215
> .

## 

Niv Sardi

Comment (Lacrymology) [inline mosto.js]: I was just answering why we got empty FPSs from caspa....
